homeassistant/components/light/__init__.py

```diff
@@ -225,6 +252,41 @@ async def async_handle_light_on_service(light, call):
 
             preprocess_turn_on_alternatives(hass, params)
 
+        supported_color_modes = light.supported_color_modes
+        # Backwards compatibility: if an RGBWW color is specified, convert to RGB + W
+        # for legacy lights
+        if ATTR_RGBW_COLOR in params:
+            legacy_supported_color_modes = light._light_internal_supported_color_modes
+            if (
+                COLOR_MODE_RGBW in legacy_supported_color_modes
+                and not supported_color_modes
+            ):
+                rgbw_color = params.pop(ATTR_RGBW_COLOR)
+                params[ATTR_RGB_COLOR] = rgbw_color[0:2]
+                params[ATTR_WHITE_VALUE] = rgbw_color[3]
+
+        # If a color is specified, convert to the color space supported by the light
+        # Backwards compatibility: Fall back to hs color if light.supported_color_modes
+        # is not implemented
+        if ATTR_HS_COLOR or ATTR_RGB_COLOR or ATTR_XY_COLOR in params:
+            if not supported_color_modes or COLOR_MODE_HS in supported_color_modes:
+                if (rgb_color := params.pop(ATTR_RGB_COLOR, None)) is not None:
+                    params[ATTR_HS_COLOR] = color_util.color_RGB_to_hs(*rgb_color)
+                if (xy_color := params.pop(ATTR_XY_COLOR, None)) is not None:
+                    params[ATTR_HS_COLOR] = color_util.color_xy_to_hs(*xy_color)
+
+            elif COLOR_MODE_RGB in supported_color_modes:
```

Comment:
Make sure every fallback has a test.

Reply:
Sure, the tests for the light component needs to be updated to test the new features as well as all the backwards compatibility cases.